src/Components/Filter/Filter.js

```diff
@@ -0,0 +1,208 @@
+import React from "react";
+import { categories } from "../../backend/db/categories";
+import { products } from "../../backend/db/products";
+import { useProduct } from "../../contexts/ProductProvider";
+
+export const getSortedData = (state, productList) => {
+  if (state.sortBy === "PRICE_HIGH_TO_LOW") {
+    return [...productList].sort((a, b) => b.price - a.price);
+  }
+  if (state.sortBy === "PRICE_LOW_TO_HIGH") {
+    return [...productList].sort((a, b) => a.price - b.price);
+  }
+  return productList;
+};
+
+export const filterByRate = (state, productList) => {
+  let data = [...productList];
+  if (state.filterBy === "1STAR_AND_ABOVE") {
+    data = [...productList].filter((product) => product.rate >= 1);
+  }
+  if (state.filterBy === "2STAR_AND_ABOVE") {
+    data = [...productList].filter((product) => product.rate >= 2);
+  }
+  if (state.filterBy === "3STAR_AND_ABOVE") {
+    data = [...productList].filter((product) => product.rate >= 3);
+  }
+  if (state.filterBy === "4STAR_AND_ABOVE") {
+    data = [...productList].filter((product) => product.rate >= 4);
+  }
+  return data;
+};
+
+export const getFilteredData = (state, productList) => {
+  let data = [...productList];
+  if (state.FilterData.filterByCategories.length !== 0) {
+    data = data.filter((product) =>
+      state.FilterData.filterByCategories.includes(product.categoryName)
+    );
+  }
+
+  return data;
+};
+
+export const getFilterByPrice = (state, productList) => {
+  let data = [...productList];
+  data = data.filter((product) => product.price >= state.getByPrice);
+  return data;
+};
+
+export function Filter() {
+  const { state, dispatch } = useProduct();
```

Comment:
just a suggestion, you can use specific dispatch name like productDispatch, cartDispatch this will make the code more specific

Reply:
Done